rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbCopyKeysNode.java

```diff
@@ -0,0 +1,101 @@
+import lombok.extern.slf4j.Slf4j;
+import org.thingsboard.common.util.JacksonUtil;
+import org.thingsboard.rule.engine.api.RuleNode;
+import org.thingsboard.rule.engine.api.TbContext;
+import org.thingsboard.rule.engine.api.TbNode;
+import org.thingsboard.rule.engine.api.TbNodeConfiguration;
+import org.thingsboard.rule.engine.api.TbNodeException;
+import org.thingsboard.rule.engine.api.util.TbNodeUtils;
+import org.thingsboard.server.common.data.plugin.ComponentType;
+import org.thingsboard.server.common.msg.TbMsg;
+import org.thingsboard.server.common.msg.TbMsgMetaData;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.regex.Pattern;
+
+@Slf4j
+@RuleNode(
+        type = ComponentType.TRANSFORMATION,
+        name = "copy keys",
+        configClazz = TbCopyKeysNodeConfiguration.class,
+        nodeDescription = "Copies the msg or metadata keys with specified key names selected in the list",
+        nodeDetails = "Will fetch fields values specified in list. If specified field is not part of msg or metadata fields it will be ignored." +
+                "If the msg is not a JSON object returns the incoming message as outbound message with <code>Failure</code> chain, " +
+                "otherwise returns transformed messages via <code>Success</code> chain",
+        uiResources = {"static/rulenode/rulenode-core-config.js"},
+        configDirective = "tbTransformationNodeCopyKeysConfig",
+        icon = "content_copy"
+)
+public class TbCopyKeysNode implements TbNode {
+
+    TbCopyKeysNodeConfiguration config;
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        this.config = TbNodeUtils.convert(configuration, TbCopyKeysNodeConfiguration.class);
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) throws ExecutionException, InterruptedException, TbNodeException {
+        Set<String> keys = config.getKeys();
+        TbMsgMetaData metaData = msg.getMetaData();
+        String msgData = msg.getData();
+        JsonNode dataNode = JacksonUtil.toJsonNode(msgData);
+        if (!dataNode.isObject()) {
+            ctx.tellFailure(msg, new RuntimeException("Msg data is not a JSON Object!"));
+            return;
+        }
+
+        if (config.isFromMetadata()) {
+            ObjectNode msgDataNode = (ObjectNode) dataNode;
+            Map<String, String> metaDataMap = metaData.getData();
+            keys.forEach(key -> {
+                Pattern pattern = Pattern.compile(key);
+                metaDataMap.forEach((keyMetaData, valueMetaData) -> {
+                    if (pattern.matcher(keyMetaData).matches()) {
+                        msgDataNode.put(keyMetaData, valueMetaData);
+                    }
+                });
+            });
+            msgData = JacksonUtil.toString(msgDataNode);
+        } else {
+            keys.forEach(key -> {
```

Comment:
I believe you should iterate through the dataNode.fields() and then check each field for matching the patterns. If it matches the pattern - `copy` it and `break` the inner cycle. In other words, no need to evaluate the same data point for multiple regexps if you already found a match for this key.